### PR TITLE
Compiler search uses a pool of workers

### PR DESCRIPTION
**Pull request by alalazo**
closes #10127

Following the request in https://github.com/spack/spack/pull/10127#pullrequestreview-186647721 I started refactoring the logic underneath:
```console
$ spack compiler add
```
The version in this PR shows much better performance than the version in `develop`. The main modifications are:

1. `spack.compilers.find_compilers` employs a `multiprocess.pool.ThreadPool` to execute system commands for the detection of compiler versions.
2. A few memoized functions have been introduced to avoid poking the filesystem multiple times for the same results.

Timing tests will be posted in the discussion below.

### CODE REVIEW

**Comment by alalazo** [comment]
**Local workstation (Ubuntu 18.04 - SSD local filesystem)**

With a1565f142c3df6549382df3c58b07d663a8ad54b:
```console
<this branch>$ time spack compiler add
==> Added 3 new compilers to /home/mculpo/.spack/linux/compilers.yaml
    gcc@8.2.0  gcc@7.3.0  clang@6.0.1-svn334776-1~exp1~20181018152737.116
==> Compilers are defined in the following files:
    /home/mculpo/.spack/linux/compilers.yaml

real    0m0,808s
user    0m0,910s
sys     0m0,289s

```
while using develop at ec2c5e5b0dfbbf5d70d0ccca91da4e8c149642fc:
```console
<develop>$ time spack compiler add
==> Added 3 new compilers to /home/mculpo/.spack/linux/compilers.yaml
    gcc@8.2.0  gcc@7.3.0  clang@6.0.1-svn334776-1~exp1~20181018152737.116
==> Compilers are defined in the following files:
    /home/mculpo/.spack/linux/compilers.yaml

real    0m1,508s
user    0m3,755s
sys     0m0,718s
```

**Fidis front-end (rhel 7.3 - GPFS filesystem)**
With a1565f142c3df6549382df3c58b07d663a8ad54b:
```console
<this-branch>$ time spack compiler add
==> Added 1 new compiler to /home/culpo/.spack/linux/compilers.yaml
    gcc@4.8.5
==> Compilers are defined in the following files:
    /home/culpo/.spack/linux/compilers.yaml

real	0m0.611s
user	0m0.432s
sys	0m0.180s
```
while using develop at ec2c5e5b0dfbbf5d70d0ccca91da4e8c149642fc:
```console
<develop>$ time spack compiler add
==> Added 1 new compiler to /home/culpo/.spack/linux/compilers.yaml
    gcc@4.8.5
==> Compilers are defined in the following files:
    /home/culpo/.spack/linux/compilers.yaml

real	0m1.908s
user	0m1.449s
sys	0m0.463s
```

**Comment by alalazo** [comment]
@tgamblin @becker33 I need to check what's going on with Travis (I get an error similar to #10186 but I can't reproduce it locally with python 2.7). Anyhow, the numbers above seem interesting - if you could give me an early feedback whether you like this approach or not I'll proceed fixing the Cray environments.

**Comment by alalazo** [comment]
The tests now pass, even though I had to skip logging some errors for Python 2.6 (see da56be6). The command:
```console
$ spack compiler add
```
is 2x-3x faster than `develop` depending on the machine. Looking forward for some early feedback.

**Comment by alalazo** [comment]
FYI, while working on this PR I noticed that Python 2.6 and Python 2.7, on `develop`, throw exceptions in the spawned processes due to unicode strings not being handled correctly. The offending part is this output captured by one of the command we use to detect the version of compilers:
```console
Command exited with status 1:
    '/usr/bin/cc' '-V'
cc: error: unrecognized command line option ‘-V’
cc: fatal error: no input files
```
where the non ascii characters `'\u2018'` and `'\u2019'` are used to report the error. With fd4e9aa I was able to solve the issue for Python 2.7, while I couldn't find a proper solution for Python 2.6 (hence da56be6).

The problem can be trimmed down to the different behavior `sys.stdout.write` exhibits between the two Python versions:
```console
$ python2.6
Python 2.6.9 (default, Apr 29 2018, 12:56:59) 
[GCC 7.3.0] on linux4
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.stdout.write(u'\u2018\n')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2018' in position 0: ordinal not in range(128)

$ python2.7
Python 2.7.15rc1 (default, Nov 12 2018, 14:31:15) 
[GCC 7.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.stdout.write(u'\u2018\n')
‘
```
This essentially means that `llnl.util.tty` currently can't handle unicode for Python 2.X. I'm not sure if this is worth a bug report or not...

File: lib/spack/llnl/util/multiproc.py [hunk 1]
```diff
@@ -8,25 +8,48 @@
...
 
 
-def spawn(f):
-    def fun(pipe, x):
-        pipe.send(f(x))
-        pipe.close()
-    return fun
+def deferred(func):
+    """Package a function call into something that can be invoked
+    at a later moment.
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/llnl/util/multiproc.py]: > I'm not sure I like this as a decorator, because it's only obvious at function definition that something is deferred. NOT at the point of invocation.

My intent was to make it clear using function names. I used `detect_version_command` instead of `detect_version` to stress that the function call was returning a command to detect the version, not the version itself. I'm not opposed to this change, just saying I'll also remove the `_command` suffix as it doesn't make much sense if we don't use `deferred` as a decorator.

File: lib/spack/spack/compiler.py [hunk 2]
```diff
@@ -286,26 +298,115 @@ def __str__(self):
                 str(self.operating_system)))))
 
 
-def _get_versioned_tuple(compiler_check_tuple):
-    full_path, prefix, suffix, detect_version = compiler_check_tuple
+@functools_backport.total_ordering
+class _CompilerID(collections.namedtuple('_CompilerIDBase', [
+    'os', 'cmp_cls', 'version'
+])):
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/compiler.py]: > If anything they make this more heavyweight... namedtuples already define total ordering and hash semantics...

So, my first try at this was just:
```python
_CompilerID = collections.namedtuple('_CompilerID', ['os', 'cmp_cls', 'version'])
```
but this clashes with the fact that we are passing around a *class* for compilers, not an instance of a class. This means we can't order `_CompilerID` because:
```python
TypeError: '<' not supported between instances of 'type' and 'type'
```
The other part is that the compiler class is unhashable, as it contains a lot of unhashable types. 

What I added here is just a workaround to be able to order on `_CompilerID` + be able to insert a `_CompilerID` into a hash based container. This is basically because uniqueness and ordering seemed important in the original algorithm used to detect compilers.

File: lib/spack/llnl/util/tty/__init__.py [hunk 3]
```diff
@@ -3,13 +3,16 @@
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 
+from __future__ import unicode_literals
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/llnl/util/tty/__init__.py]: > I'm curious whether #10186 solves the root problem, though -- if it does, maybe we don't need this import everywhere?

Unfortunately rebasing on `develop` and checking if #10186 fixed this issue was my first try and it didn't work :sob: Using 3b8b13809e4f9e8d7456c4d2e27d7ff069c8e44e:
```console
$ python2.6 bin/spack -d python
==> Reading config file /home/mculpo/PycharmProjects/spack/etc/spack/defaults/modules.yaml
==> Reading config file /home/mculpo/PycharmProjects/spack/etc/spack/defaults/linux/modules.yaml
==> Reading config file /home/mculpo/PycharmProjects/spack/etc/spack/defaults/config.yaml
==> Reading config file /home/mculpo/.spack/linux/config.yaml
Spack version 0.12.0
Python 2.6.9, Linux x86_64
>>> import llnl.util.tty as tty
>>> tty.info(u'Some unicode string')
==> Some unicode string
>>> tty.info(u'Some unicode string\u2018')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/mculpo/PycharmProjects/spack/lib/spack/llnl/util/tty/__init__.py", line 91, in info
    cprint("@%s{%s==>} %s" % (format, st_text, cescape(str(message))),
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2018' in position 19: ordinal not in range(128)

$ python2.7 bin/spack -d python
==> Reading config file /home/mculpo/PycharmProjects/spack/etc/spack/defaults/modules.yaml
==> Reading config file /home/mculpo/PycharmProjects/spack/etc/spack/defaults/linux/modules.yaml
==> Reading config file /home/mculpo/PycharmProjects/spack/etc/spack/defaults/config.yaml
==> Reading config file /home/mculpo/.spack/linux/config.yaml
Spack version 0.12.0
Python 2.7.15rc1, Linux x86_64
>>> import llnl.util.tty as tty
>>> tty.info(u'Some unicode string')
==> Some unicode string
>>> tty.info(u'Some unicode string\u2018')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/mculpo/PycharmProjects/spack/lib/spack/llnl/util/tty/__init__.py", line 91, in info
    cprint("@%s{%s==>} %s" % (format, st_text, cescape(str(message))),
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2018' in position 19: ordinal not in range(128)
```
one can see that `tty` fails if we try to print some non-ascii character in Python 2. Here I have been able to solve the issue for Python 2.7 but for Python 2.6 I found no solution - the error seems in the implementation of `sys.stdout.write`, see https://github.com/spack/spack/pull/10190#issuecomment-449962219. I didn't dig deeper, but given that Python 2.6 suport has been discontinued a long time ago it might be a bug in the implementation of `sys.stdout`.

File: lib/spack/spack/compiler.py [hunk 4]
```diff
@@ -286,26 +298,115 @@ def __str__(self):
...
+
+    compilers_s = []
+    for (compiler_id, name_variation, lang), (return_value, error) \
+            in zip(tags, compiler_versions):
+        # If we had an error, move to the next element
+        if error:
+            try:
+                # This will fail on Python 2.6 if a non ascii
+                # character is in the error
+                tty.debug(error)
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/compiler.py]: Yes, see above.

File: lib/spack/llnl/util/multiproc.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tgamblin) [reply lib/spack/llnl/util/multiproc.py]: Yeah, I saw that.  I think it's still clearer to write defer(f)(*args) -- the `_command` convention still wasn't clear to me and I had to go look up the function to make sure I knew what I was reading.

[the remark this answers is not in the document]
Comment (tgamblin) [reply lib/spack/llnl/util/multiproc.py]: Yeah, I saw that.  I think it's still clearer to write defer(f)(*args) -- the `_command` convention still wasn't clear to me and I had to go look up the function to make sure I knew what I was reading.

File: lib/spack/llnl/util/tty/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tgamblin) [reply lib/spack/llnl/util/tty/__init__.py]: ok I think it's a reasonable fix, then.

File: lib/spack/llnl/util/filesystem.py [hunk 5]
```diff
@@ -1283,3 +1283,37 @@ def find_libraries(libraries, root, shared=True, recursive=False):
     libraries = ['{0}.{1}'.format(lib, suffix) for lib in libraries]
 
     return LibraryList(find(root, libraries, recursive))
+
+
+@memoized
+def is_accessible_dir(path):
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/llnl/util/filesystem.py]: Taken care of in 27d7549

File: lib/spack/llnl/util/multiproc.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/llnl/util/multiproc.py]: Fixed in e72f8ebd11abcd9f0ba027c2709a10d949629411

File: lib/spack/llnl/util/multiproc.py [hunk 6]
```diff
@@ -8,25 +8,48 @@
...
+            return func(*args, **kwargs)
+        return _deferred_call
+    return _impl
+
+
+def invoke(f):
+    return f()
+
+
+def execute(command_list, map_fn=map, transformation=invoke):
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/llnl/util/multiproc.py]: > Maybe I'm just used to map, but I think it's a decent abstraction... and a more familiar one. 

Agreed. With `execute` I tried to make the business logic more evident at call site by hiding the deferred call - couldn't do much though to hide the `mapping` function. Since the gain is any case small (and not without contention) I think it's right to remove it.

[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/llnl/util/multiproc.py]: > Why can't we assemble a matrix of arguments then pass it to functions rather than deferring all the calls? I like defer, but I think separating the data from the calls separates the concerns more, especially if we want to be portable between a thread pool and a process pool -- in the latter case you have to pickle arguments.

**TL;DR** I'll look into how hard is to do that. I mean, conceptually it's not that difficult, but if we need to use a `ProcessPool` I need to check which arguments are involved and whether they can be pickled or not. 

---

Right now we are basically farming out system calls. Before implementing the PR, I checked this small example on a 4 cores laptop:
```console
Python 3.6.5 (default, Apr  1 2018, 05:46:30) 
[GCC 7.3.0] on linux

>>> import multiprocessing.pool
>>> import subprocess 
>>> import timeit
>>>
>>> def foo():
...     p = subprocess.Popen(['sleep', '1'])
...     p.communicate()
...     return 'Done'
>>> tp = multiprocessing.pool.ThreadPool()
>>> l = [foo] * 8
>>> timeit.timeit('print(tp.map(lambda x: x(), a))', globals={'a': a, 'tp': tp}, number=2)
['Done', 'Done', 'Done', 'Done', 'Done', 'Done', 'Done', 'Done']
['Done', 'Done', 'Done', 'Done', 'Done', 'Done', 'Done', 'Done']
4.041111436999927
>>>
>>> timeit.timeit('print(list(map(lambda x: x(), a)))', globals={'a': a}, number=2)
['Done', 'Done', 'Done', 'Done', 'Done', 'Done', 'Done', 'Done']
['Done', 'Done', 'Done', 'Done', 'Done', 'Done', 'Done', 'Done']
16.08182063300046
```
to be confident that a `ThreadPool` could parallelize this part. Any line of Python code we add to `foo` I think it will be effectively serialized by a `ThreadPool` because of the GIL. If the overhead of that part is non negligible we'll need to move to a `ProcessPool` - and that's where the funny `pickle` part starts.

File: lib/spack/spack/compiler.py [hunk 7]
```diff
@@ -223,52 +228,59 @@ def fc_version(cls, fc):
...
+        """Returns a list of commands that, when invoked, search for compilers
+        in the paths supplied.
+
+        Args:
+            operating_system (OperatingSystem): the OS requesting the search
+            *search_paths (list of paths): paths where to look for a
+                compiler
+
+        Returns:
+            (tags, commands): ``tags`` is a list of compiler tags, containing
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/compiler.py]: That was just because of defensive programming. The data structure used internally before was a list and order was said to be important in a couple of comments:

https://github.com/spack/spack/blob/9f799957180ddf9e710fb3386a23ee824f56b9aa/lib/spack/spack/compiler.py#L267-L270

As you ask for it, I'll look into simplifying the code using a `dict`.

File: lib/spack/spack/compiler.py [hunk 8]
```diff
@@ -286,26 +298,115 @@ def __str__(self):
...
-        if (not version) or (not str(version).strip()):
-            tty.debug(
-                "Couldn't get version for compiler %s" % full_path)
-            return None
-        return (version, prefix, suffix, full_path)
-    except ProcessError as e:
-        tty.debug(
-            "Couldn't get version for compiler %s" % full_path, e)
-        return None
+        version = callback(path)
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/compiler.py]: I'll try to work towards this state of things, see also my reply above https://github.com/spack/spack/pull/10190#discussion_r245974635. 

> That would make the code more portable between thread pools [...]. Basically don't pass code as arguments -- make the args to mapped functions pure data.

I think it's more complicated than that. Python functions can be pickled (see [here](https://docs.python.org/3/library/pickle.html#what-can-be-pickled-and-unpickled)) if they are defined at the top level of a module - or better *their fully qualified name is pickled*. According to what I read in the docs we might have troubles unpickling instances of classes with e.g. class defined attributes (like compilers). In any case I'll check what can be done there.

File: lib/spack/spack/compiler.py [hunk 9]
```diff
@@ -223,52 +228,59 @@ def fc_version(cls, fc):
...
-                for pre, name, suf in prod:
-                    regex = r'^(%s)%s(%s)$' % (pre, re.escape(name), suf)
+        files_to_be_tested = llnl.util.filesystem.files_in(*search_paths)
+
+        tags, commands = [], []
+        for language in ('cc', 'cxx', 'f77', 'fc'):
+
+            # Get compiler names and the callback to detect their versions
+            compiler_names = getattr(cls, '{0}_names'.format(language))
+            callback = getattr(cls, '{0}_version'.format(language))
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/compiler.py]: See reply above.

File: lib/spack/spack/compiler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/compiler.py]: Giving a better look right now at the stuff in `spack.compilers` module. I see that in fact we have all the functions we need there. I can remove all the complexity and just pass the compiler name. Then I can use:
```python
compiler_cls = spack.compilers.class_for_compiler_name(name)
```
in client code.

File: lib/spack/spack/compiler.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/compiler.py]: Done in 5448ef0

File: lib/spack/spack/compiler.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/compiler.py]: Done in 5448ef0

File: lib/spack/spack/compiler.py [hunk 10]
```diff
@@ -286,26 +298,115 @@ def __str__(self):
...
+        compilers_s.append((tag, path))
+
+    compilers_s.sort()
+
+    compilers_d = {}
+    for sort_key, group in itertools.groupby(compilers_s, key=lambda x: x[0]):
+        compiler_id, name_variation, language = sort_key
+        by_compiler_id = compilers_d.setdefault(compiler_id, {})
+        by_name_variation = by_compiler_id.setdefault(name_variation, {})
+        by_name_variation[language] = list(x[1] for x in group).pop()
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/compiler.py]: Done in 5448ef0

File: lib/spack/spack/compiler.py [hunk 11]
```diff
@@ -286,26 +298,115 @@ def __str__(self):
...
+        if version == 'unknown':
+            continue
+
+        tag = compiler_id._replace(version=version), name_variation, lang
+        compilers_s.append((tag, path))
+
+    compilers_s.sort()
+
+    compilers_d = {}
+    for sort_key, group in itertools.groupby(compilers_s, key=lambda x: x[0]):
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/compiler.py]: Done in 5448ef0

File: lib/spack/llnl/util/multiproc.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/llnl/util/multiproc.py]: Done in 5448ef0. I stayed with a `ThreadPool` for now, but tried a `Pool` with Python 3.6 and it works.

File: lib/spack/spack/compiler.py [hunk 12]
```diff
@@ -223,52 +228,59 @@ def fc_version(cls, fc):
...
+        in the paths supplied.
+
+        Args:
+            operating_system (OperatingSystem): the OS requesting the search
+            *search_paths (list of paths): paths where to look for a
+                compiler
+
+        Returns:
+            (tags, commands): ``tags`` is a list of compiler tags, containing
+                all the information on a compiler, but version. ``commands``
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/compiler.py]: Changed in 5448ef0, so not relevant anymore.

File: lib/spack/llnl/util/multiproc.py [hunk 13]
```diff
@@ -8,25 +8,48 @@
 than multiprocessing.Pool.apply() can.  For example, apply() will fail
 to pickle functions if they're passed indirectly as parameters.
 """
-from multiprocessing import Process, Pipe, Semaphore, Value
+import functools
+from multiprocessing import Semaphore, Value
 
-__all__ = ['spawn', 'parmap', 'Barrier']
+__all__ = ['Barrier']
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/llnl/util/multiproc.py]: There are no other methods as of 5448ef0

File: lib/spack/spack/compiler.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/compiler.py]: Done in 5448ef0

File: lib/spack/spack/compilers/__init__.py [hunk 14]
```diff
@@ -166,17 +168,27 @@ def all_compiler_specs(scope=None, init_config=True):
...
+    Returns:
+        List of compilers found in the supplied paths
     """
-    # Find compilers for each operating system class
-    oss = all_os_classes()
-    compiler_lists = []
-    for o in oss:
-        compiler_lists.extend(o.find_compilers(*paths))
-    return compiler_lists
+    tags, commands = [], []
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/compilers/__init__.py]: Now it's a single list. I hope it could be acceptable as a solution.

**Comment by alalazo** [comment]
@tgamblin If the code in `spack/compilers/__init__.py` looks good enough I'll proceed restoring the two unit tests under `FIXME` and porting the Cray part. In case, let me know if the code needs further explanations in the comments or docstrings.

**Comment by alalazo** [comment]
Real time reported for compiler detection as of 397cdf3:


PR   | Ubuntu 18.04 (SSD ext4) | CentOS 7.3 (GPFS) | Cray (Piz Daint) |
------ | --------------------------------- | --------------------------- | --------------------- |
*Before this PR* | 0m1,442s | 0m1.862s | 0m1.577s |
*After this PR* | 0m0,680s | 0m0.650s | 0m1.520s |

Summary: the time to detect compilers for usual linux systems improved substantially, Cray remained quite the same (not surprising as the main optimizations were on avoiding useless `os.stat` and the Cray detection system is based modules instead of on file searching).

**Comment by alalazo** [comment]
@tgamblin @becker33 I think this is ready for another review.

File: lib/spack/spack/compilers/__init__.py [hunk 15]
```diff
@@ -368,6 +412,196 @@ def all_compiler_types():
...
+#: - variation: a NameVariation for file being tested
+#: - language: compiler language being tested (one of 'cc', 'cxx', 'fc', 'f77')
+#: - path: full path to the executable being tested
+#:
+DetectVersionArgs = collections.namedtuple(
+    'DetectVersionArgs', ['id', 'variation', 'language', 'path']
+)
+
+
+def arguments_to_detect_version_fn(operating_system, paths, override=True):
```
Comment (alalazo) [inline lib/spack/spack/compilers/__init__.py]: @tgamblin @becker33 I moved all the functions that are needed to search for compilers in this module, as I think they belong here as free functions and not in `spack.architecture.OperatingSystem` as methods.  I had to leave room for overriding when porting Cray compilers, as that's currently the only architecture that needs a special treatment. 

I am not particularly proud of the `override=True` in the arguments above, that is needed to re-use the `_default` implementation for Cray front-end, and will probably spend some time looking for alternative approaches that leave the default behavior as a nested function that is not visible externally. Any idea is welcome.

Comment (tgamblin) [reply lib/spack/spack/compilers/__init__.py]: @alalazo: I don't like it either -- if you can figure out how best to get rid of it that would be awesome.

Comment (alalazo) [reply lib/spack/spack/compilers/__init__.py]: I had a better look at the code and the part where I need to override the function was basically to compute different search paths. Therefore I added another point of customization for the OS in aa702a0. The result looks cleaner to me than before. 

If the PR is fine with you in the current state I'll write some docs on how to add a new compiler and ping you back. Let me know if I should proceed with that.

File: lib/spack/spack/compilers/__init__.py [hunk 16]
```diff
@@ -176,18 +185,53 @@ def all_compiler_specs(scope=None, init_config=True):
...
+    # of arguments for each call.
+    arguments = []
+    for o in all_os_classes():
+        search_paths = getattr(o, 'compiler_search_paths', default_paths)
+        arguments.extend(arguments_to_detect_version_fn(o, search_paths))
+
+    # Here we map the function arguments to the corresponding calls
+    tp = multiprocessing.pool.ThreadPool()
+    detected_versions = tp.map(detect_version, arguments)
+    tp.close()
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/compilers/__init__.py]: It should indeed. Originally it was in a context manager, but I had to abandon that as it's not supported for Python 2.

File: lib/spack/llnl/util/filesystem.py [hunk 17]
```diff
@@ -1351,3 +1352,64 @@ def find_libraries(libraries, root, shared=True, recursive=False):
...
+    Args:
+        *path_hints (list of paths): list of paths taken into
+            consideration for a search
+
+    Returns:
+        A list containing the real path of every existing directory
+        in `path_hints` and its `bin` subdirectory if it exists.
+    """
+    # Select the realpath of existing directories
+    existing_paths = filter(os.path.isdir, map(os.path.realpath, path_hints))
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/llnl/util/filesystem.py]: I simplified the function as suggested below, see 2d5aadc

File: lib/spack/llnl/util/filesystem.py [hunk 18]
```diff
@@ -1351,3 +1352,64 @@ def find_libraries(libraries, root, shared=True, recursive=False):
...
+
+    # Adding their 'bin' subdirectory
+    def maybe_add_bin(path):
+        bin_subdirectory = os.path.realpath(os.path.join(path, 'bin'))
+        if os.path.isdir(bin_subdirectory):
+            return [path, bin_subdirectory]
+        return [path]
+
+    return list(
+        itertools.chain.from_iterable(map(maybe_add_bin, existing_paths))
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/llnl/util/filesystem.py]: Done in 2d5aadc

File: lib/spack/spack/compilers/__init__.py [hunk 19]
```diff
@@ -176,18 +185,53 @@ def all_compiler_specs(scope=None, init_config=True):
...
+        List of compilers found
     """
-    # Find compilers for each operating system class
-    oss = all_os_classes()
-    compiler_lists = []
-    for o in oss:
-        compiler_lists.extend(o.find_compilers(*paths))
-    return compiler_lists
+    # Turn the path hints into real paths that are to be searched
+    path_hints = path_hints or get_path('PATH')
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/compilers/__init__.py]: Done in ea7910a

File: lib/spack/spack/compilers/__init__.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/compilers/__init__.py]: Done

File: lib/spack/spack/test/cmd/release_jobs.py [hunk 20]
```diff
@@ -36,6 +39,11 @@ def test_specs_deps(tmpdir, config):
     validate(deps_object, specs_deps_schema)
 
 
+@pytest.mark.skipif(
+    sys.version_info[:2] < (2, 7),
+    reason="For some reason in Python2.6 we get a utf-32 string "
+           "that can't be parsed"
+)
```
[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/test/cmd/release_jobs.py]: Done

[the remark this answers is not in the document]
Comment (alalazo) [reply lib/spack/spack/test/cmd/release_jobs.py]: I removed this change, because this test leaves fixtures etc. in a bad state when failing. See https://travis-ci.org/spack/spack/jobs/542632380 We probably need to refactor it. I'll open an issue so that we don't forget.

**Comment by alalazo** [comment]
Ready for further review!

File: lib/spack/spack/compilers/__init__.py [hunk 21]
```diff
@@ -176,18 +185,55 @@ def all_compiler_specs(scope=None, init_config=True):
...
+    Returns:
+        List of compilers found
     """
-    # Find compilers for each operating system class
-    oss = all_os_classes()
-    compiler_lists = []
-    for o in oss:
-        compiler_lists.extend(o.find_compilers(*paths))
-    return compiler_lists
+    if path_hints is None:
```
Comment (scheibelp) [inline lib/spack/spack/compilers/__init__.py]: @alalazo I think this check might be overly strict. I'm seeing cases where `path_hints` is `[]` for `spack compiler find` (which I expect would check `PATH`) and so this check is not triggered.
